crates/pint-query/src/main.rs

```diff
@@ -0,0 +1,152 @@
+        }
+        (None, Some(key_hex)) => key_hex.0,
+        (None, None) => anyhow::bail!("At least one of key name or key hex should be provided."),
+    };
+
+    let node_client = EssentialNodeClient::new(node_address)?;
+    let output = node_client
+        .query_state(contract_address.to_owned(), query_key)
+        .await?;
+    println!("{}", serde_json::to_string(&output)?);
+    Ok(())
+}
+
+// FIXME: Should be made obsolete by https://github.com/essential-contributions/essential-base/issues/228
+#[derive(Clone, Debug)]
+struct Key(essential_types::Key);
+
+impl FromStr for Key {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Ok(Self(
+            hex::decode(s)?
+                .chunks_exact(8)
+                .map(|chunk| word_from_bytes(chunk.try_into().expect("Always 8 bytes")))
+                .collect(),
+        ))
+    }
+}
+
+/// Find the file within the current directory or parent directories with the given name.
+fn find_file(mut dir: PathBuf, file_name: &str) -> Option<PathBuf> {
+    loop {
+        let path = dir.join(file_name);
+        if path.exists() {
+            return Some(path);
+        }
+        if !dir.pop() {
+            return None;
+        }
+    }
+}
+
+/// Given a `ManifestFile`, return the `ContractABI` of the already compiled contract.
+fn get_contract_abi(manifest: ManifestFile) -> anyhow::Result<ContractABI> {
+    let out_dir = manifest.out_dir();
+    for entry in read_dir(out_dir.clone())? {
+        let entry = entry?;
+        let entry_file_name = entry.file_name();
+        let name = entry_file_name
+            .to_str()
+            .expect("file name should be convertible to string");
+        if name.ends_with("-abi.json") {
+            return pint_abi::from_path(&entry.path()).map_err(|err| anyhow::anyhow!("{}", err));
+        }
+    }
+    Err(anyhow::anyhow!(
+        "Could not find *-abi.json file in {:?}",
+        out_dir
+    ))
+}
+
+/// Given a `ContractABI` and a key name, return the `Key`.
+fn get_key_from_abi(abi: ContractABI, key_name: String) -> anyhow::Result<essential_types::Key> {
+    for (index, storage) in abi.storage.iter().enumerate() {
+        if storage.name == key_name {
+            if storage.ty == TypeABI::Bool
+                || storage.ty == TypeABI::Int
+                || storage.ty == TypeABI::Real
+                || storage.ty == TypeABI::String
+                || storage.ty == TypeABI::B256
+            {
+                let key = vec![index.try_into()?];
+                return Ok(key);
+            } else {
+                // FIXME: support complex types
+                return Err(anyhow::anyhow!(
+                    "Querying key of type {:?} with name is not supported. Try providing a key_hex.",
+                    storage.ty
+                ));
+            }
+        }
+    }
+    Err(anyhow::anyhow!(
+        "Could not find key {key_name} in given ABI"
+    ))
+}
```

Comment:
What about something more functional like (this is untested):
```
/// Given a `ContractABI` and a key name, return the `Key`.
fn get_key_from_abi(abi: ContractABI, key_name: String) -> anyhow::Result<essential_types::Key> {
    abi.storage
        .iter()
        .enumerate()
        .find(|(_, storage)| storage.name == key_name)
        .map(|(index, storage)| {
            if matches!(
                storage.ty,
                TypeABI::Bool | TypeABI::Int | TypeABI::Real | TypeABI::String | TypeABI::B256
            ) {
                let key = vec![index.try_into()?];
                Ok(key)
            } else {
                // FIXME: support complex types
                Err(anyhow::anyhow!(
                    "Querying key of type {:?} with name {key_name} is not supported. \
                     Try providing a key_hex.",
                    storage.ty
                ))
            }
        })
        .unwrap_or_else(|| {
            Err(anyhow::anyhow!(
                "Could not find key {key_name} in given ABI"
            ))
        })
}
```

Reply:
This may not work due to the `?` operator.. so this probably needs some work.. or just leave the `for` loops.